### ISSUE TITLE
Crash to desktop when highlighting in TextFields that contain certain strings

### ISSUE THREAD

**Issue by DesignDefect** [openfl/openfl#2602]
**Describe the bug**
Highlighting to the end of a TextField containing 'ff', 'fl', or 'fi' and using NotoSans Regular as the font causes  Linux HXCPP and HashLink targets to crash. 

**To Reproduce**
[Example.zip](https://github.com/openfl/openfl/files/9881322/Example.zip)
Modified AddingText example. Font file included in zipped project.
```haxe
package;

import openfl.display.Sprite;
import openfl.text.Font;
import openfl.text.TextField;
import openfl.text.TextFormat;
import openfl.Assets;

class Main extends Sprite {
	public function new () {
		super ();
		var format = new TextFormat (Assets.getFont("assets/NotoSans-Regular.ttf").fontName, 30, 0x7A0026);
		var textField = new TextField ();
		textField.defaultTextFormat = format;
		textField.embedFonts = true;
		textField.selectable = true;
		textField.x = 50;
		textField.y = 50;
		textField.width = 200;
		textField.text = "fl";
		addChild (textField);
	}
}
```
1. `openfl test linux`
2. Click and highlight to the end of the text field
3. Crash to desktop, error below

```bash
Called from hxcpp::__hxcpp_main
Called from ApplicationMain::main ApplicationMain.hx line 26
Called from ApplicationMain::create ApplicationMain.hx line 135
Called from lime.app.Application::exec lime/app/Application.hx line 150
Called from lime._internal.backend.native.NativeApplication::exec lime/_internal/backend/native/NativeApplication.hx line 146
Called from lime._internal.backend.native.NativeApplication::handleRenderEvent lime/_internal/backend/native/NativeApplication.hx line 371
Called from lime.app._Event_lime_graphics_RenderContext_Void::dispatch lime/_internal/macros/EventMacro.hx line 91
Called from openfl.display.Stage::__onLimeRender openfl/display/Stage.hx line 2106
Called from openfl.display.OpenGLRenderer::__render openfl/display/OpenGLRenderer.hx line 764
Called from openfl.display.OpenGLRenderer::__renderDrawable openfl/display/OpenGLRenderer.hx line 850
Called from openfl.display._internal.Context3DDisplayObjectContainer::renderDrawable openfl/display/_internal/Context3DDisplayObjectContainer.hx line 30
Called from openfl.display.OpenGLRenderer::__renderDrawable openfl/display/OpenGLRenderer.hx line 850
Called from openfl.display._internal.Context3DDisplayObjectContainer::renderDrawable openfl/display/_internal/Context3DDisplayObjectContainer.hx line 30
Called from openfl.display.OpenGLRenderer::__renderDrawable openfl/display/OpenGLRenderer.hx line 850
Called from openfl.display._internal.Context3DDisplayObjectContainer::renderDrawable openfl/display/_internal/Context3DDisplayObjectContainer.hx line 30
Called from openfl.display.OpenGLRenderer::__renderDrawable openfl/display/OpenGLRenderer.hx line 858
Called from openfl.display._internal.Context3DTextField::renderDrawable openfl/display/_internal/Context3DTextField.hx line 39
Called from openfl.display._internal.Context3DTextField::render openfl/display/_internal/Context3DTextField.hx line 24
Called from openfl.display._internal.CairoTextField::render openfl/display/_internal/CairoTextField.hx line 390
Called from lime.graphics.cairo.Cairo::showGlyphs lime/graphics/cairo/Cairo.hx line 370
Error : Invalid object
```
**Expected behavior**
Selected text should be highlighted

**OpenFL Targets**
Linux, Hashlink

**Additional context**
NotoSans appears to have ligatures defined for these character pairs.
![image](https://user-images.githubusercontent.com/44974839/198350853-39c6608c-36b0-47d6-8652-4c69f9a09956.png)
TextField selection logic does not appear to recognize these correctly. When used in a longer string the highlight logic appears to be off by one after the ff ligature.
![image](https://user-images.githubusercontent.com/44974839/198349633-a53fa6d1-4036-4bee-9c74-3dadcea00015.png)

**Comment by tobil4sk** [comment]
Reproduces on Windows as well. This is because when rendering a text that has a selection, the render code doesn't take ligatures into account.

For example, if I have a 5 character string which contains a ligature for ff, these two characters combine into one glyph, and so the 5 characters map onto 4 glyphs. The selection index and caret index of the text field is based on the character positions, so the start and end indexes are 0 and 4 if the whole text is selected for example: https://github.com/openfl/openfl/blob/332cb95f2531b0a0a8f92e537e8057edc84298a6/src/openfl/display/_internal/CairoTextField.hx#L344-L345

The error happens when it attempts to use the characters indexes on the glyphs array. When using the character index 4, it goes out of bounds since there are only 4 glyphs (indexes 0-3) and no mapping is performed to adjust the character index to the glyph index.

https://github.com/openfl/openfl/blob/332cb95f2531b0a0a8f92e537e8057edc84298a6/src/openfl/display/_internal/CairoTextField.hx#L389

Therefore, the out-of-bounds read of the glyphs array returns `null` and the crash comes later when trying to read the properties of this null glyph.

**Comment by tobil4sk** [comment]
Here is a patch that prevent the crash, although it doesn't really address the root cause of the issue: https://github.com/tobil4sk/openfl/commit/aed53c18bf19d0d47c8ad80ebaf71e4a137011c3

**Comment by joshtynjala** [comment]
> Here is a patch that prevent the crash, although it doesn't really address the root cause of the issue

I just applied a modified version of this patch that takes into account more complex text with multiple groups.

**Comment by tobil4sk** [comment]
Hm, I've only just switched over to your version of the patch, it doesn't seem to prevent the crash from my testing.

**Comment by player-03** [comment]
Can confirm that joshtynjala's patch doesn't work, unless you select from the very beginning of the text (`selectionStart == 0`).

I think part of the problem was the check happened too soon. Extra arithmetic was required to try to account for text groups, but since it was so far away from the array access, it was harder to notice that the arithmetic was wrong.

tobil4sk's solution is much closer to being correct, it just needs to happen _after_ subtracting out `group.startIndex`. This also puts it right next to the array access, making it much easier to verify.

```diff
selectionStart -= group.startIndex;
selectionEnd -= group.startIndex;
+if (selectionEnd > glyphs.length)
+{
+	selectionEnd = glyphs.length;
+}
+
for (i in selectionStart...selectionEnd)
	selectedGylphs.push(glyphs[i]);
```

**Comment by dimensionscape** [comment]
I'm not sure either approach is correct, but its in the right direction.

**Comment by player-03** [comment]
Agreed, we need to address the root issue. The above is still just a stopgap.
